fix: retain inventory for pruning fail/skip

sdowell opened the pull request:
Prior to this change, the inventory always was deleted at the end of a Destroy event. This would occur even in the case of a pruning skip/failure, resulting in the objects being removed from the inventory without being deleted. This change makes it so that the inventory is only deleted if all objects have been deleted and reconciled.

sdowell commented:
/assign @karlkfi

karlkfi commented on pkg/apply/task/inv_set_task.go:
```diff
@@ -54,79 +57,114 @@ func (i *InvSetTask) Identifiers() object.ObjMetadataSet {
+// prunes were failed or skipped, the inventory will be updated.
+//
+// If Destroy is false, the inventory will be updated.
+func (i *DeleteOrUpdateInvTask) Start(taskContext *taskrunner.TaskContext) {
```
i usually like to put public methods above private methods, since it's more likely someone came to the file to find them. interface before impl.

sdowell replied:
done

karlkfi commented on pkg/apply/task/inv_set_task.go:
```diff
@@ -54,79 +57,114 @@ func (i *InvSetTask) Identifiers() object.ObjMetadataSet {
+		if i.Destroy {
+			skippedDeletes := taskContext.InventoryManager().SkippedDeletes()
+			failedDeletes := taskContext.InventoryManager().FailedDeletes()
+			// if any deletes were failed/skipped, retain the inventory
```
I suspect we should also keep the inventory if any reconciles failed.

This would slightly change the behavior of the ConfigSync finalizer, but the fact that CS ignores reconciler failures is a CS impl detail, not something we want to do in cli-utils or kpt.

In fact, because the CS finalizer is a preview annotation, we can still change how it works to wait for reconciles, even if the applier doesn't.

karlkfi replied:
i think you can also clean up the code here a little by extracting a `success() bool` so you can just do `if destroy && !success() { err = deleteInventory() } else { err = updateInventory() }; ch <- result(err); return`

sdowell replied:
Done. Refactored the code and added reconcile failed/skipped check to the success condition

karlkfi commented on pkg/apply/solver/solver.go:
```diff
@@ -211,26 +211,22 @@ func (t *TaskQueueBuilder) Build(taskContext *taskrunner.TaskContext, o Options)
-		})
+	prevInvIds, _ := t.InvClient.GetClusterObjs(t.invInfo)
+	klog.V(2).Infoln("adding delete/update inventory task")
+	var setInventoryTaskName string
```
nit: just `taskName`

sdowell replied:
done

karlkfi reviewed (changes requested):
good start!

needs some e2e tests validating that destroy can fail, be retried, and succeed.

sdowell commented:
@karlkfi Added an e2e test and updated the logic to not consider delete/reconcile skips. I also added timeoutReconciles as a condition to consider a failure

karlkfi commented on pkg/apply/task/inv_set_task.go:
```diff
@@ -54,79 +83,99 @@ func (i *InvSetTask) Identifiers() object.ObjMetadataSet {
 
-// StatusUpdate is not supported by the InvSetTask.
-func (i *InvSetTask) StatusUpdate(_ *taskrunner.TaskContext, _ object.ObjMetadata) {}
+func (i *DeleteOrUpdateInvTask) success(taskContext *taskrunner.TaskContext) bool {
```
func comment: success returns true when destroy actuation and reconciliation was fully successful. When true, it's safe to delete the inventory.

the name could maybe also be destroySuccessful, since it doesn't work for checking apply success.

sdowell replied:
done

karlkfi commented on test/e2e/e2e_test.go:
```diff
@@ -191,6 +191,10 @@ var _ = Describe("E2E", func() {
 					pruneRetrieveErrorTest(ctx, c, invConfig, inventoryName, namespace.GetName())
 				})
 
+				It("PruneReconcileError", func() {
```
rename DestroyReconciliationFailure

sdowell replied:
done

karlkfi commented on test/e2e/e2eutil/common.go:
```diff
@@ -198,6 +210,21 @@ func AssertUnstructuredExists(ctx context.Context, c client.Client, obj *unstruc
+	}, resultObj)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred(),
+		"expected GET not to error (%s): %s", ref, err)
+	gomega.Expect(resultObj.GetDeletionTimestamp()).NotTo(gomega.BeNil(),
```
i dont know if deletion timestamp is always nil after parsing. there's an IsZero func I think. The k8s timestamp is a weird wrapper.

sdowell replied:
done

karlkfi commented on pkg/apply/solver/solver.go:
```diff
@@ -211,26 +211,22 @@ func (t *TaskQueueBuilder) Build(taskContext *taskrunner.TaskContext, o Options)
+	klog.V(2).Infoln("adding delete/update inventory task")
+	var taskName string
+	if o.Destroy {
+		taskName = "delete-inventory-0"
```
lets change the test name of this task too. It will require some find/replace, but I want to make it clear that it's not always a delete, because it shows up in logs. how about "inventory-delete-or-update-0"?

sdowell replied:
done

karlkfi commented:
/ok-to-test

karlkfi commented on test/e2e/e2eutil/common.go:
```diff
@@ -198,6 +210,21 @@ func AssertUnstructuredExists(ctx context.Context, c client.Client, obj *unstruc
+	gomega.Expect(err).NotTo(gomega.HaveOccurred(),
+		"expected GET not to error (%s): %s", ref, err)
+	gomega.Expect(resultObj.GetDeletionTimestamp().IsZero()).To(gomega.BeFalse(),
+		"expected deletion timestamp to be non-nil (%s)", ref)
```
to be non-zero

sdowell replied:
done

karlkfi commented on pkg/apply/solver/solver.go:
```diff
@@ -211,34 +211,30 @@ func (t *TaskQueueBuilder) Build(taskContext *taskrunner.TaskContext, o Options)
 // to the cluster. Returns a pointer to the Builder to chain function calls.
 func (t *TaskQueueBuilder) newApplyTask(applyObjs object.UnstructuredSet,
-	applyFilters []filter.ValidationFilter, applyMutators []mutator.Interface, o Options) taskrunner.Task {
+		applyFilters []filter.ValidationFilter, applyMutators []mutator.Interface, o Options) taskrunner.Task {
```
is this a format on save thing? I don't know if a go formatting standard was chosen for this project, but this looks inconsistent.

karlkfi replied:
maybe a go version change in fmt?

sdowell replied:
yeah probably my IDE being opinionated

sdowell replied:
reverted

karlkfi reviewed (approved):
/lgtm

sdowell commented:
/retest

sdowell commented:
`Job execution failed: Pod got deleted unexpectedly`
/retest

karlkfi commented:
/retest

sdowell commented:
/retest

sdowell commented:
/retest

sdowell commented:
/retest

sdowell commented:
/retest

sdowell commented:
/retest

sdowell commented:
/test cli-utils-presubmit-master-stress

karlkfi commented:
/retest

sdowell commented:
/test cli-utils-presubmit-master-stress

sdowell commented:
/test cli-utils-presubmit-master-stress

sdowell commented:
/test cli-utils-presubmit-master-stress

karlkfi commented:
/lgtm
